wire-runtime/src/main/java/com/squareup/wire/MessageAdapter.java

```diff
@@ -534,6 +558,29 @@ private void writeValueNoTag(WireOutput output, Object value, Datatype datatype)
 
   // Reading
 
+  /** Read an encoded message from {@code source}. */
+  public M read(BufferedSource source) throws IOException {
+    checkNotNull(source, "source == null");
+    return read(new WireInput(source));
+  }
+
+  /** Read an encoded message from {@code bytes}. */
+  public M readBytes(byte[] bytes) throws IOException {
+    checkNotNull(bytes, "bytes == null");
+    Buffer buffer = new Buffer();
+    M message = read(buffer.write(bytes));
+    if (!buffer.exhausted()) {
+      throw new IllegalArgumentException("");
+    }
+    return message;
+  }
+
+  /** Read an encoded message from {@code stream}. */
+  public M readStream(InputStream stream) throws IOException {
```

Comment:
* Gson requires a `Reader` or `JsonReader`.
 * Moshi requires a `BufferedSource` or `JsonReader`.
 * Wire 1.x was broken in the same way as this PR.
 * Protobuf documents it consumes the whole `InputStream` or requires you to limit or use `CodedInputStream`.
 * Jackson appears to be broken in the same way but I stopped following the ridiculously abstracted code after a while to verify.